<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd"><html xmlns="http://www.w3.org/1999/xhtml"><head><title>R: Golub ExpressionSet</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>

<table width="100%" summary="page for esGolub {NMF}"><tr><td>esGolub {NMF}</td><td style="text-align: right;">R Documentation</td></tr></table>

<h2>Golub ExpressionSet</h2>

<h3>Description</h3>

<p>This data comes originally from the gene expression data
from <cite>Golub et al. (1999)</cite>. The version included in
the package is the one used and referenced in
<cite>Brunet et al. (2004)</cite>. The samples are from 27
patients with acute lymphoblastic leukemia (ALL) and 11
patients with acute myeloid leukemia (AML).
</p>


<h3>Format</h3>

<p>There are 3 covariates listed.
</p>

<ul>
<li><p> Samples: The original sample labels.  </p>
</li>
<li><p> ALL.AML: Whether the
patient had AML or ALL. It is a <code><a href="../../base/html/factor.html">factor</a></code> with levels
<code>c('ALL', 'AML')</code>.  </p>
</li>
<li><p> Cell: ALL arises from two different types of
lymphocytes (T-cell and B-cell).  This specifies which for the ALL patients;
There is no such information for the AML samples. It is a
<code><a href="../../base/html/factor.html">factor</a></code> with levels <code>c('T-cell', 'B-cell', NA)</code>.
</p>
</li></ul>


<h3>Details</h3>

<p>The samples were assayed using Affymetrix Hgu6800 chips
and the original data on the expression of 7129 genes
(Affymetrix probes) are available on the Broad Institute
web site (see references below).
</p>
<p>The data in <code>esGolub</code> were obtained from the web
page related to the paper from <cite>Brunet et al.
(2004)</cite>, which describes an application of Nonnegative
Matrix Factorization to gene expression clustering. (see
link in section <em>Source</em>).
</p>
<p>They contain the 5,000 most highly varying genes
according to their coefficient of variation, and were
installed in an object of class
<code><a href="../../Biobase/html/ExpressionSet-class.html">ExpressionSet-class</a></code>.
</p>


<h3>Source</h3>

<p>Web page for <cite>Brunet2004</cite>:<br />
<a href="http://www.broadinstitute.org/publications/broad872">http://www.broadinstitute.org/publications/broad872</a>
</p>
<p>Original data from Golub et al.:<br />
<code>http://www-genome.wi.mit.edu/mpr/data_set_ALL_AML.html</code>
</p>


<h3>References</h3>

<p>Golub TR, Slonim DK, Tamayo P, Huard C, Gaasenbeek M,
Mesirov JP, Coller H, Loh ML, Downing JR, Caligiuri Ma,
Bloomfield CD and Lander ES (1999). &quot;Molecular
classification of cancer: class discovery and class
prediction by gene expression monitoring.&quot; _Science (New
York, N.Y.)_, *286*(5439), pp. 531-7. ISSN 0036-8075,
&lt;URL: http://www.ncbi.nlm.nih.gov/pubmed/10521349&gt;.
</p>
<p>Brunet J, Tamayo P, Golub TR and Mesirov JP (2004).
&quot;Metagenes and molecular pattern discovery using matrix
factorization.&quot; _Proceedings of the National Academy of
Sciences of the United States of America_, *101*(12), pp.
4164-9. ISSN 0027-8424, &lt;URL:
http://dx.doi.org/10.1073/pnas.0308531101&gt;, &lt;URL:
http://www.ncbi.nlm.nih.gov/pubmed/15016911&gt;.
</p>


<h3>Examples</h3>

<pre>


# requires package Biobase to be installed
if( require(Biobase) ){

	data(esGolub)
	esGolub
	## Not run: pData(esGolub)

}
</pre>

<hr /><div style="text-align: center;">[Package <em>NMF</em> version 0.21.0 <a href="00Index.html">Index</a>]</div>
</body></html>
